codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/Synthetic.java

```diff
@@ -55,6 +54,25 @@ protected Node createNode() {
         throw new CodegenException("attempted to serialize runtime only trait");
     }
 
+    @Override
+    public boolean equals(Object other) {
+        if (other == this) {
+            return true;
+        } else if (!(other instanceof Trait)) {
+            return false;
+        } else if (hashCode() != other.hashCode()) { // take advantage of hashcode caching
```

Comment:
It does not look like the hash code is being cached below, and this should also test that the class is `Synthetic` instead of just `Trait`. Having said that and given that there's no really a good way to tell any instance apart I'd make them all equal or all different on its reference value (I suspect that the later will be preferred in particular when trying to figure out if this has been processed). E.g.,

Making each instance equivalent:

```
public boolean equal(Object other) {
   if (other == this) {
    return true;
  }
  return (other instanceof Synthetic);
}

public int hashCode() {
    return toShapeId().hashCode();
}
```
Making each instance different:

```
public boolean equal(Object other) {
   return other == this;
}

public int hashCode() {
    return System.identityHashCode(this);
}
```